### ISSUE TITLE
Fold stream.timepoint.join when timepoints are transitively reached.

### ISSUE THREAD

**Issue by benvanik** [iree-org/iree#7822]
We should look through the use-def chain to track which timepoints are already reached and redirect them into `stream.timepoint.immediate` values (that then fold nicely).

For example:
```mlir
    %454 = stream.cmd.execute await(%result_timepoint_78) => with(...) {
        ...
    } => !stream.timepoint
    %455 = stream.resource.dealloca await(%454) => %result_77 : !stream.resource<transient>{%c2360576} => !stream.timepoint
    %456 = stream.timepoint.join max(%455, %454) => !stream.timepoint
```
Here it's clear from the use-def chain that if timepoint %455 is reached then %454 must have also been reached and it could be rewritten as:
```mlir
    %454 = stream.cmd.execute await(%result_timepoint_78) => with(...) {
        ...
    } => !stream.timepoint
    %455 = stream.resource.dealloca await(%454) => %result_77 : !stream.resource<transient>{%c2360576} => !stream.timepoint
    %satisfied454 = stream.timepoint.immediate
    %456 = stream.timepoint.join max(%455, %satisfied454) => !stream.timepoint
```
->
```mlir
    %454 = stream.cmd.execute await(%result_timepoint_78) => with(...) {
        ...
    } => !stream.timepoint
    %455 = stream.resource.dealloca await(%454) => %result_77 : !stream.resource<transient>{%c2360576} => !stream.timepoint
    %456 = %455  // from join folder
```

(this happens on the alloca awaits as well)
To start we could just do this within a must-be-executed scope (block and all post-dominated blocks). It should happen after the timepoint propagation pass runs - though we may then want to run that pass again.

**Comment by benvanik** [comment]
There is also the case with `stream.timepoint.await`:
```mlir
    %308 = stream.resource.dealloca await(%307) => %result_57 : !stream.resource<transient>{%c1049856} => !stream.timepoint
    %309 = stream.timepoint.join max(%308, %307) => !stream.timepoint
    %310:2 = stream.timepoint.await %309 => %304, %305 : !stream.resource<staging>{%c4}, !stream.resource<transient>{%c401408}
    ....
    %result_59, %result_timepoint_60 = stream.resource.alloca uninitialized : !stream.resource<transient>{%c2622720} => !stream.timepoint
    %316 = stream.timepoint.join max(%result_timepoint_60, %308, %307) => !stream.timepoint
    %317 = stream.cmd.execute await(%316) 
```
Here we know that because %309 is waited on any execution after that point must have %307 and %308 signaled.

**Comment by benvanik** [comment]
This should probably be done in a pass with dfx - in programs with loops it's possible for resolved timepoints to escape across block/function boundaries. This could be an analysis on the timepoint SSA values that tracks the ops that may guarantee they are signaled and then per op that consumes them a query for if one of those signaling ops is in the must-be-executed set.

**Comment by benvanik** [comment]
Fixed in #9754. #9747 still needs work but a majority of the timepoints go away with the elision pass.
